fvm/state/address_generator_state.go

```diff
@@ -0,0 +1,67 @@
+package state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const keyAddressState = "account_address_state"
+
+type AddressGeneratorState struct {
+	ledger Ledger
+	chain  flow.Chain
+}
+
+func NewAddressGeneratorState(ledger Ledger, chain flow.Chain) *AddressGeneratorState {
+	return &AddressGeneratorState{
+		ledger: ledger,
+		chain:  chain,
+	}
+}
+
+func (state *AddressGeneratorState) Init() {
+	state.Set(state.chain.NewAddressGenerator())
+}
+
+func (state *AddressGeneratorState) GetGenerator() (flow.AddressGenerator, error) {
+	stateBytes, err := state.ledger.Get("", "", keyAddressState)
+	if err != nil {
+		return nil, err
+	}
+
+	addressGenerator := state.chain.BytesToAddressGenerator(stateBytes)
+
+	return &stateBoundAddressGenerator{
+		addressGenerator:      addressGenerator,
+		addressGeneratorState: state,
+	}, nil
+}
+
+func (state *AddressGeneratorState) Set(generator flow.AddressGenerator) {
+	stateBytes := generator.Bytes()
+	state.ledger.Set("", "", keyAddressState, stateBytes)
+}
+
+// stateBoundAddressGenerator is a decorator for an address generator.
+// It uses the underlying generator.
+// The only change is that when next address is called the state is updated as well
+type stateBoundAddressGenerator struct {
```

Comment:
This is great! However, I'm wondering if it's necessary to have both `AddressGeneratorState` and `stateBoundAddressGenerator`. Could they be combined into one struct that implements `flow.AddressGenerator`, or is it useful to have a separate struct for state management?

Reply:
The extra complexity from the AGState is totally not needed, thanks for pointing that out. I'll remove it!

Reply:
@psiemens I reworked this part. Let me know what you think.